xwiki-platform-core/xwiki-platform-oldcore/src/main/java/com/xpn/xwiki/internal/doc/XWikiAttachmentList.java

```diff
@@ -0,0 +1,239 @@
+ *
+ * This software is distributed in the hope that it will be useful,
+ * but WITHOUT ANY WARRANTY; without even the implied warranty of
+ * MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the GNU
+ * Lesser General Public License for more details.
+ *
+ * You should have received a copy of the GNU Lesser General Public
+ * License along with this software; if not, write to the Free
+ * Software Foundation, Inc., 51 Franklin St, Fifth Floor, Boston, MA
+ * 02110-1301 USA, or see the FSF site: http://www.fsf.org.
+ */
+package com.xpn.xwiki.internal.doc;
+
+import java.util.ArrayList;
+
+import java.util.Collection;
+import java.util.ListIterator;
+import java.util.List;
+import java.util.Iterator;
+import java.util.Map;
+import java.util.concurrent.ConcurrentSkipListMap;
+
+import com.xpn.xwiki.doc.XWikiAttachment;
+import com.xpn.xwiki.doc.XWikiDocument;
+
+import com.xpn.xwiki.internal.AbstractNotifyOnUpdateList;
+
+/**
+ * AttachmentList that holds elements in order of filename.
+ * 
+ * @version $Id$
+ * @param <E>
+ * @since 10.0RC1
+ */
+public class XWikiAttachmentList extends AbstractNotifyOnUpdateList<XWikiAttachment>
+{
+
+    private final Map<String, XWikiAttachment> map = new ConcurrentSkipListMap<String, XWikiAttachment>();
+
+    private final XWikiDocument document;
+
+    /**
+     * Initializes the map.
+     * 
+     * @since 10.0RC1
+     */
+    public XWikiAttachmentList(XWikiDocument document)
+    {
+        super(new ArrayList<XWikiAttachment>());
+        this.document = document;
+    }
+
+    /**
+     * Adds attachment to the list in order of filename.
+     * 
+     * @param attachment XWikiAttachment to add to the list
+     * @since 10.0RC1
+     */
+    @Override
+    public boolean add(XWikiAttachment attachment)
+    {
+        map.put(attachment.getFilename(), attachment);
+        this.list = new ArrayList<XWikiAttachment>(map.values());
+        onUpdate();
+        added(attachment);
+        return true;
+    }
+
+    /**
+     * Adds attachment to the list in order of filename.
+     * 
+     * @param index index is ignored as list is reordered based on filename
+     * @param attachment XWikiAttachment to add to the list
+     * @since 10.0RC1
+     */
+    @Override
+    public void add(int index, XWikiAttachment attachment)
+    {
+        add(attachment);
+    }
+
+    /**
+     * {@inheritDoc}
+     * 
+     * @since 10.0RC1
+     */
+    @Override
+    public void clear()
+    {
+        super.clear();
+        map.clear();
+    }
+
+    /**
+     * {@inheritDoc}
+     * 
+     * @since 10.0RC1
+     */
+    @Override
+    public boolean addAll(Collection<? extends XWikiAttachment> c)
+    {
+        for (XWikiAttachment x : c) {
+            map.put(x.getFilename(), x);
+            added(x);
+        }
+        this.list = new ArrayList<XWikiAttachment>(map.values());
+        onUpdate();
+        return true;
+    }
+
+    /**
+     * {@inheritDoc}
```

Comment:
You should probably copy/paste the explanation you have in add(int index, XWikiAttachment attachment) to be clear.